### DISCUSSION TITLE
CDKサンプルにて作成されるインフラについてシステム構成図を自動生成してほしい

### DISCUSSION THREAD

**Discussion by koichimurakami** [mbc-net/mbc-cqrs-serverless#44] [General]
MBC CQRS サーバーレス フレームワークで作成するインフラは作成するリソース数が多いため全体像を把握することが困難です。
もし可能であればCDKにて作成するリソースをシステム構成図として自動作成してもらえないでしょうか？

**Comment by baquan1708** [comment]
@koichimurakami 
 
After research, I found 2 libraries that support genreating the infrastructure diagram from CDK: [cdk-dia ](https://github.com/pistazie/cdk-dia)and [pdk](https://aws.github.io/aws-pdk/developer_guides/cdk-graph-plugin-diagram/index.html) (pdk is inspired by cdk-dia). I've prepared two versions of it. Which diagram version should I use?

<figure>
  <figcaption style="text-align:center;">1)This diagram is genreated by [cdk-dia].</figcaption>
  <img src="https://github.com/user-attachments/assets/698c9a6b-978e-4943-9318-227d7522e690" alt="diagram-cdk-dia">
</figure>

<figure>
  <figcaption style="text-align:center;">2)This diagram is genreated by [pdk].</figcaption>
  <img src="https://github.com/user-attachments/assets/aa6ef4b3-05fe-4686-bb8f-d27c2d5688dc" alt="diagram-pdk">
</figure>

**Reply by koichimurakami** [reply]
@baquan1708 
aws-pdk is well maintenaced recently. That's why I prefer aws pdk.
How do you think?

**Reply by baquan1708** [reply]
@koichimurakami 
I agree with you. Additionally, the library intends to add more features that would be helpful.

I will proceed to implement this feature now.

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero

I've just create a PR for this feature: https://github.com/mbc-net/mbc-cqrs-serverless/pull/47. Could you please take a look?

**Reply by Tuan-Anh-Nguyen-pionero** [reply]
@baquan1708 cc @koichimurakami
Murakami-san would like to upload the infrastructure diagram to Artifact. Could you implement it? 
Also would  like to upload the Swagger file. If you have any other suggestions, we can discuss them together here.

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 
I'm able to implement this, but I’m still unclear on the purpose of uploading the infrastructure diagram to the artifact after running CodeBuild in the CodePipeline flow (given that the artifacts are stored in an Amazon S3 bucket).

If the reasoning is, for example, to share the diagram among developers or to track changes for each deployment. Storing the diagram in Git seems like a suitable alternative.

**Reply by koichimurakami** [reply]
@baquan1708 @Tuan-Anh-Nguyen-pionero 
Managing changes using Git
Is the pipeline in an infinite loop?
Or are we worried that the deployment will fail?
I would like to avoid changing to Git if there are no changes to the API or infrastructure. 
Is that possible?

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 
Actually, What I meant is that the infrastructure diagram and API definition are generated locally on the developer’s computer. These changes are then committed to Git, which triggers the CodePipeline. This way, the pipeline only starts after developers intentionally push their updates to Git. 

This way, there is not an infinite loop. The API Swagger file and infrastructure diagram are generated only when the local process is successful.

![image](https://github.com/user-attachments/assets/dae65a5d-f014-4a4f-81eb-c5a4b755251e)

**Reply by koichimurakami** [reply]
@baquan1708 cc: @Tuan-Anh-Nguyen-pionero 
Developers must be conscious of generating documents, and there is a possibility of making mistakes.

Some people do not update documents manually, and there may be cases where documents cannot be generated depending on the development environment such as Windows, Mac, or Linux.

That's why I want automatic generation logic.
I think it's good to commit to Git manually and do it within the development flow, but I think it would be even better if the framework had automatic generation logic.

You can download it from artifacts and commit it to Git.

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 
Thank you very much for your detailed response. I now fully understand the issue, and I will proceed with implementing the automatic generation logic as discussed. The infrastructure diagram and API documentation will be auto generated and they will be uploaded to artifact.

**Reply by baquan1708** [reply]
@koichimurakami cc: @Tuan-Anh-Nguyen-pionero 
I've created the PR https://github.com/mbc-net/mbc-cqrs-serverless/pull/61. This PR auto generate diagram and store it in artifact. Could you please take a look?

I will proceed with implementing the API documentation generation in the pipeline.

**Comment by koichimurakami** [comment]
Release Info in Japanese: https://www.mbc-net.com/update-mbc-cqrs-%e3%82%b5%e3%83%bc%e3%83%90%e3%83%bc%e3%83%ac%e3%82%b9-%e3%83%95%e3%83%ac%e3%83%bc%e3%83%a0%e3%83%af%e3%83%bc%e3%82%af-ver-0-1-32-beta-cdk%e3%81%a7%e3%83%87%e3%83%97%e3%83%ad/
